### PR TITLE
Enable custom usertext-instance in usertext smarty plugin

### PR DESCRIPTION
**Pull request by alexispeter**
- Adds an extensible base for list-based usertext filters with wildcard capabillities, uses logic that was previously implemented on the badwords-paging
- Allows usage of pre-configured usertext instances in usertext smarty plugin for special filter usage

### CODE REVIEW

**Comment by alexispeter** [comment]
please review @fauvel

This approach now only applies the filter once, as opposed to applying the filter multiple times, which would not work with the spamword-highlighting filter.
As a result, the badword-filter will no longer replace badwords that are created by application of the filter itself (see changes to the unit-test), which is probably not necessary since its highly unlikely that this would ever happen.

File: library/CM/Usertext/Filter/ListReplace.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+    /**
+     * @param $text
+     * @return string
+     */
+    public function replaceMatch($text) {
+        $text = (string) $text;
+        $regex = $this->_toRegex();
+        $count = preg_match_all($regex, $text);
+        if ($count) {
```
Comment (fauvel) [inline library/CM/Usertext/Filter/ListReplace.php]: Why this first step?

File: library/CM/Usertext/Filter/ListReplace.php [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    /**
+     * @param $text
+     * @return string
+     */
+    public function replaceMatch($text) {
+        $text = (string) $text;
+        $regex = $this->_toRegex();
+        $count = preg_match_all($regex, $text);
```
Comment (fauvel) [inline library/CM/Usertext/Filter/ListReplace.php]: Why this first step?

File: library/CM/Usertext/Filter/ListReplace.php [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    /**
+     * @param string $phrase
+     * @return string
+     */
+    protected function _transformItemToRegex($phrase) {
+        $regexp = preg_quote($phrase, '#');
+        $regexp = str_replace('\*', '[^A-Za-z]*', $regexp);
+        $regexp = str_replace('\|', '\b', $regexp);
+        $regexp = '\S*' . $regexp . '\S*';
+        return $regexp;
```
Comment (fauvel) [inline library/CM/Usertext/Filter/ListReplace.php]: I would add the delimiters & flags `#...#i` right here (they belong to the regex)

File: library/CM/Usertext/Filter/ListReplace.php [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    /**
+     * @return string
+     */
+    private function _toRegex() {
+        $cacheKey = $this->_getCacheKey();
+        $cache = CM_Cache_Shared::getInstance();
+        $phrasesRegex = null;
+        if (false == ($phrasesRegex = $cache->get($cacheKey))) {
```
Comment (fauvel) [inline library/CM/Usertext/Filter/ListReplace.php]: Use `===`

File: library/CM/Usertext/Filter/ListReplace.php [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        return CM_CacheConst::Usertext_Filter_ReplaceList . '_phrasesHash:' . md5(implode('', iterator_to_array($this->_phrases)));
+    }
+
+    /**
+     * @return string
+     */
+    private function _toRegex() {
+        $cacheKey = $this->_getCacheKey();
+        $cache = CM_Cache_Shared::getInstance();
+        $phrasesRegex = null;
```
Comment (fauvel) [inline library/CM/Usertext/Filter/ListReplace.php]: You don't need to initialize here

File: library/CM/Usertext/Filter/ListReplace.php (hunk 2 above)
Comment (alexispeter) [inline library/CM/Usertext/Filter/ListReplace.php]: code from an earlier approach I forgot to take out

File: library/CM/Usertext/Filter/ListReplace.php [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+        return false;
+    }
+
+    /**
+     * @param string $userInput
+     * @return bool
+     */
+    public function isMatch($userInput) {
+        if (preg_match($this->_toRegex(), (string) $userInput)) {
```
Comment (fauvel) [inline library/CM/Usertext/Filter/ListReplace.php]: What about `return (bool) preg_match()`?

File: tests/library/CM/Usertext/Filter/BadwordsTest.php [hunk 7]
```diff
@@ -2,61 +2,21 @@
...
+        $this->assertSame('… bar', $filter->transform('foo bar', new CM_Frontend_Render()));
+    }
 
-        $actual = $filter->transform("hello foobar there", $render);
-        $this->assertSame("hello ${replace} there", $actual);
-        $actual = $filter->transform("hello XfoobarX there", $render);
-        $this->assertSame("hello ${replace} there", $actual);
-        $actual = $filter->transform("hello mayo.foobar.ran there", $render);
-        $this->assertSame("hello ${replace} there", $actual);
+    public function testTransformCustomtReplacement() {
```
Comment (fauvel) [inline tests/library/CM/Usertext/Filter/BadwordsTest.php]: Typo (`testTransformCustomReplacement `)

File: library/CM/Usertext/Filter/ListReplace.php [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+     */
+    public function isMatch($userInput) {
+        return (boolean) preg_match($this->_toRegex(), (string) $userInput);
+    }
+
+    /**
+     * @return string
+     */
+    private function _getCacheKey() {
+        return CM_CacheConst::Usertext_Filter_ReplaceList . '_phrasesHash:' . md5(implode('', iterator_to_array($this->_phrases)));
```
Comment (fauvel) [inline library/CM/Usertext/Filter/ListReplace.php]: Is that caching!? Computing the cache key is almost as expensive as computing the cached value...

Comment (alexispeter) [inline library/CM/Usertext/Filter/ListReplace.php]: Is caching even needed?

Comment (fauvel) [inline library/CM/Usertext/Filter/ListReplace.php]: Caching the source paging for sure.

Comment (alexispeter) [inline library/CM/Usertext/Filter/ListReplace.php]: Did a quick benchmark to be sure, calculating the regex takes much less time than the actual cache-call.

File: library/CM/Usertext/Filter/Badwords.php [hunk 9]
```diff
@@ -1,12 +1,12 @@
...
-    public function transform($text, CM_Frontend_Render $render) {
-        $text = (string) $text;
-        $badwordList = new CM_Paging_ContentList_Badwords;
-        $text = $badwordList->replaceMatch($text, '…');
-
-        return $text;
+    /**
+     * @param string|null $replace
+     */
+    public function __construct($replace = null) {
```
Comment (fauvel) [inline library/CM/Usertext/Filter/Badwords.php]: Maybe rename to `$replacementPattern` for clarity?

**Comment by fauvel** [comment]
lgtm!

**Pull request merged**